### ISSUE TITLE
Behavior of operator>> should more closely resemble that of built-in overloads.

### ISSUE THREAD

**Issue by TurpentineDistillery** [nlohmann/json#367]
Basically:
* eat up leading blanks or newlines
* leave the stream where the parsing ended, allowing for more data downstream, which may or may not be json. 

```c++
#include "json.hpp"
    
template<typename T>
static void test(const std::string& s)
{
    T val{999};
    std::stringstream sstr(s);
    //sstr.exceptions(std::ifstream::failbit | std::ifstream::badbit);

    try {
        size_t i = 0;
        while(sstr >> val) {
            std::cerr << "    result[" << i++ << "]=" << val 
                      << "; streampos=" << sstr.tellg() << std::endl;
        }   

        std::cerr << "    final values: result=" << val 
                  << " i=" << i
                  << " streampos=" << sstr.tellg()
                  << " streamflags=" << sstr.flags() 
                  << std::endl;
    } catch(std::exception& e) {
        std::cerr << "    std::exception:" << e.what() << std::endl;
    }   

    std::cerr << std::endl;
}

int main()
{
    using json = nlohmann::json;
    size_t i = 0;

    for(std::string s : { 
        "", 
        "   ",
        "111",
        "222 \t\n",
        "\n\t 333", 
        " 111 \n222\n \n  333" })
    {   
        std::cerr << "\nTest-case[" << i++ << "]: '" << s << "'" << std::endl;
    
        std::cerr << "With int:" << std::endl;
        test<int>(s);

        std::cerr << "With json:" << std::endl;
        test<json>(s);

        std::cerr << "----" << std::endl;
    }   

    return 0;
}
```

```
Test-case[0]: ''
With int:
    final values: result=999 i=0 streampos=-1 streamflags=4098

With json:
    std::exception:parse error - unexpected end of input

----

Test-case[1]: '   '
With int:
    final values: result=999 i=0 streampos=-1 streamflags=4098

With json:
    std::exception:parse error - unexpected end of input

----

Test-case[2]: '111'
With int:
    result[0]=111; streampos=-1
    final values: result=111 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=111; streampos=-1
    std::exception:basic_string::append

----

Test-case[3]: '222 	
'
With int:
    result[0]=222; streampos=3
    final values: result=222 i=1 streampos=-1 streamflags=4098

With json:
    final values: result=222 i=0 streampos=-1 streamflags=4098

----

Test-case[4]: '
	 333'
With int:
    result[0]=333; streampos=-1
    final values: result=333 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=333; streampos=-1
    std::exception:basic_string::append

----

Test-case[5]: ' 111 
222
 
  333'
With int:
    result[0]=111; streampos=4
    result[1]=222; streampos=9
    result[2]=333; streampos=-1
    final values: result=333 i=3 streampos=-1 streamflags=4098

With json:
    std::exception:parse error - unexpected number literal; expected end of input

----
```

**Comment by nlohmann** [comment]
The lines `std::exception:basic_string::append` revealed a bug in the `fill_line_buffer` function where

```
m_line_buffer.append(n - 1, '\x01');
```

was executed with `n = 0`. This must be fixed.

About the inputs:

1. `""` (empty input): An empty input is not a valid JSON document. Hence, I think the message `parse error - unexpected end of input` is adequate.
2. `"   "` (white space): After ignoring whitespace, this is again empty input, cf. input 1.
3. `"111"` (number): Parsing this input yields a number. Parsing the (then empty) input again, is the same as input 1.
4. `"222 \t\n"` (number with trailing whitespace): The input is parsed as a number while reading the trailing whitespace. Parsing the (then empty) input again, is the same as input 1.
5. `"\n\t 333"` (number with leading whitespace): Same as input 4.
6. `" 111 \n222\n \n  333"` (number with leading whitespace, followed by additional numbers): A number must not be followed by a number, so the message `parse error - unexpected number literal; expected end of input` makes sense to me (note that after merging #301 this message will contain the byte offset of the error).

I can understand your thoughts about the behavior of the streams. However, I think it is not a good idea to stop parsing once we found a JSON value, because this would mean that code like

```cpp
json j;
j << "false foo bar";
```

would run without exception.

Am I too strict about this?

**Comment by TurpentineDistillery** [comment]
> An empty input is not a valid JSON document.

An empty input is not a valid integer either, yet cin>>i will not throw - it will consume the blanks and leave i unchanged.

There are two classes of use-cases:
* Parse a singular json document from a string (e.g. using parse(...) method or a free function). Here the library should verify that the input is not empty and that there's nothing past the end.
* Parse next json document from a stream (operator>>). Here the code cannot assume that the stream must contain a singleton json document and nothing else. Maybe there's some binary data that follows - it's not our business to know.

Think of a what kind of code a new user of the library would write, assuming they have not read any documentation other some examples, when they try to read a bunch of json values from stdin in a loop. Now make the API work that way, following the "design principle of least astonishment" : )

**Comment by nlohmann** [comment]
I understand your point and you answered my "Am I too strict about this?" question with "yes" 😄 

It would be nice to hear other opinions about this. I still feel uncomfortable accepting non-valid JSON files by just looking at a prefix...

**Comment by gregmarr** [comment]
Seems reasonable for streaming to stop after a JSON document has been read.  If this is a behavior change, it might require a major version bump.

Note that both `j << cin;` and `cin >> j;` call the same code.

I agree that this seems weird, but I'm not sure it's valid right now, since there is no operator<< that takes a string:

    json j;
    j << "false foo bar";

**Comment by TurpentineDistillery** [comment]
Note for the future:
Suppose input stream contains arbitrarily large, or infinite number of json records (e.g. sensor data being piped from some upstream process), which happen to not be newline-delimited. calling getline() on such stream will fail.

**Comment by krzysztofwos** [comment]
Allowing to partially parse a stream would be very useful. In fact, that's what I'm struggling with right now. I get input from a WebSocket, and it's basically a stream of non-delimited JSON objects, e.g. `{...}{...}`. If parsing stopped once an object is found, this code would be valid:

```
json msg;
while (iss >> msg) {
    // ...
}
```
Right now I can't see any way to parse this kind of stream.

**Comment by mrkgnao** [comment]
+1, this is basically the only thing I wish for here. Is there any way to simulate this behavior with the library as it stands now?

**Comment by nlohmann** [comment]
Unfortunately not.

**Comment by gregmarr** [comment]
I saw someone do it in a way that required a two-stage parse.  They parsed the data once, got the position of the parse error, assumed that was the start of the next whole entry, and so parsed again limiting to just before the parse error, and then moving the start position to the error position for the next entry.

**Comment by krzysztofwos** [comment]
That's exactly what I did some time ago. Try to parse, catch the exception, save the offset, limit stream to that many characters, parse, repeat. I think that required the `feature-userexceptions` branch and of course it's not terribly efficient.

**Comment by nlohmann** [comment]
This is currently the problem in the preparation of 3.0.0 - a lot of API-changing features depend on each other, and I want to try to avoid releasing a 4.0.0 and 5.0.0 in a short period.

**Comment by gregmarr** [comment]
@krzysztofwos Yeah, I think it was you that I was thinking of, but I couldn't find the post on slack, I think it's beyond the 10K limit.

**Comment by mrkgnao** [comment]
I just resorted to signaling the end of a JSON message with `\0`. It'll work for now. :)

**Comment by TurpentineDistillery** [comment]
If you're in control of both producing and consuming json records, another approach is to dump in single-sine mode, and, when consuming, first fetch the json document from stream with `getline`, and then parse.

**Comment by nlohmann** [comment]
The current parser relies on reading input streams with `getline()`. This makes it quite hard to parse several JSON values in a row, because in values like `true false\n"hello"`, `false` is lost, because it was read with `true`, but never processed to a value. I think I need to use the underlying `std::basic_streambuf`.

**Comment by szikra** [comment]
Will be support added for parsing multiple JSON objects from strings too? 
(Which I was also asking for in #438)

**Comment by gregmarr** [comment]
@szikra That's exactly what this issue is about.  operator>> uses parse().

**Comment by szikra** [comment]
@gregmarr Thanks.
The comment "This class is sort of depreciated or at least you are normally told to use other classes." is not very reassuring though. :)

I'm sure my `stringstream<<string` is not the most efficient way to create a stream from a buffer, it was just a test, but at least it's not as ugly as other suggestions on stackoverflow :) And most of them are "implementation defined".
`stringstream<<char*` is also an accepted answer http://stackoverflow.com/questions/5188140/c-const-char-to-stringstream

And yes, it would be enough if we had a public method that returns the length parse() read the last time.
Or maybe parse could return the processed length too.
Something like this:
```
class J{
public:
	struct C {
		J* j;
		int len;
		operator J&() {
			return *j;
		}
	};

	C parse() {
		return C {this, 123};
	}
};
```

RVO would help here to discard length if we are not using it, so it wouldn't have to be stored as last processed amount inside json, right?
The downside is, that this would break `auto res=j.parse();`  (`J res=j.parse();` should continue to work.)

**Comment by gregmarr** [comment]
Yeah, that first comment isn't terribly useful.  The next one is better, or the one linked in a comment:
http://stackoverflow.com/questions/2079912/simpler-way-to-create-a-c-memorystream-from-char-size-t-without-copying-t
I think an optional pointer-to-something argument to `parse()` that receives a length (safer) or end pointer would be better than changing the return type.

**Comment by szikra** [comment]
Yeah, I don't want to break existing code. I also hate returning data with pointer arguments, so ugly. :)
Adding an extended parse method should fix the 'auto' issue
```
class J{
public:
	struct C {
		J* j;
		int len;
		operator J&() {
			return *j;
		}
	};

	C parseWithLen() {
		return C {this, 123};
	}
	J& parse() {
		return parseWithLen();
	}
       
	/// But if people really want, this could also be added as an option:
	J& parse(size_t& len) {
		C c = parseWithLen();
		len = c.len;
		return *c.j;
	}
	J& parse(size_t* len) {
		C c = parseWithLen();
		if (len) *len = c.len;
		return *c.j;
	}

};
```

**Comment by TurpentineDistillery** [comment]
There are some subtleties to this.
The partial-parse is unambiguous in case of objects, arrays, strings, and literals, but not for numbers. Suppose the input stream contains `000123....` The syntactically correct repeated parsing would yield a sequence of json-records: 0, 0, 0, 123, (...) since an integer can be 0, but can't have leading zeros. However, this would likely be semantically wrong.  Proper behavior would be to parse "000123" as a group of digits, and then error-out saying it's not a valid json rather than yield a sequence of syntactically-correct/semantically-likely-wrong records.

**Comment by nlohmann** [comment]
@TurpentineDistillery With the fix for #452 the library would indeed reject `000123` as ill-formed number.

**Comment by TurpentineDistillery** [comment]
@nlohmann 
For partial stream parsing we need to guarantee that the parser/lexer shall not consume the stream past the end of json record (with caveats mentioned above). This does not bode well with the "black-box" re2c approach. How do you feel about dispensing with re2c-generated parser it and replacing with a hand-written lexer/parser, provided that it meets the coding standards of this library (i.e. correct, efficient, and understandable)? This would also make the library "actually" single-header : )

**Comment by nlohmann** [comment]
I think it is doable with re2c - the current state is due to the fact that I originally did not think about parsing from a stream, let alone partial parsing. A hand-written lexer may have advantages, but I am currently afraid of the complexity of this. (Maybe if I fail to adjust re2c, I have different mindset...)

About the single-headerness: I think it's not a problem so far, and @theodelrieu already proposed in #474 to split the code into several files for development and merge then for release. :)

**Comment by TurpentineDistillery** [comment]
I tried sticking with re2c but didn't like where I was going with that, so I took a stab at hand-coding a parser and it was way simpler than I thought it would be. Having a custom parser that works off of a template `input_iterator` would also address the issues related to contiguous-iterator requirement.
So just FYI, if there are difficulties with sticking with re2c, there's a plan-B.

**Comment by gregmarr** [comment]
There's been a whole thread on the std-proposals google group about what kind of things would be required for standardization of a JSON library.  One big thing that's being discussed now is using SAX vs StAX for streaming.  I have absolutely no experience with either of these things, and don't even really know what they are, just wanted to pass the info along.

**Comment by nlohmann** [comment]
@TurpentineDistillery Good to hear that there is a plan B :-)

@gregmarr I am aware of this discussion. With the callback function we can pass to our parser, I think we somehow support a SAX-like approach. I'm not sure though how SAX evangelists would think about this :-)

Looking at this discussion, I was surprised how emotional people can be over parsing. Whatever we do, we should work on the marketing of the library in this regard. It seems as if people would reject this library if the "right" words are missing in the README.

**Comment by nlohmann** [comment]
Sorry about my confusion: is there a different in the semantics of `operator<<` and `operator>>`?

**Comment by TurpentineDistillery** [comment]
In the spirit of emulating the standard library, `operator<<` would be used to serialize to stream and `operator>>` would be used to deserialize from stream, and leave it at that, whereas the library implements these operators both ways. No harm with leaving things as they are, or, perhaps, deprecating `j << istream` / `j >> ostream`

**Comment by nlohmann** [comment]
Thanks for the feedback! Deprecating these functions is a good idea.

**Comment by nlohmann** [comment]
I need some advice regarding the parser.

I am currently experimenting with a hand-written lexer that works without re2c. As a side effect, I rewrote the buffering of input streams and I now can "partially" parse streams, effectively allowing code like

```cpp
std::stringstream ss;
ss << "[1,2,3] false";
json j1, j2;
j1 << ss; // j1 == [1,2,3]
j2 << ss; // j2 == false
```

That is, an input stream is not necessarily read until eof.

While this behavior (not forcing to read the entire input) may make sense for `operator<<`, it would feel weird for the `parse` functions. For instance, I would expect the following code to throw a parse error, because not the entire string has been processed:

```cpp
json j = json::parse("[1,2,3] false");
```

So I would propose the following behavior:

- All flavors of `parse` assume the given input is a single JSON text that should be deserialized. That is, when the parser finishes, there is at most whitespace left until eof.
- The `operator<<` has the semantics described above, so there are no more `expected end of input` errors.

I hope you could follow me... I would be happy for comments, because there are a lot of errors to fix in the feature branch ;)

**Comment by krzysztofwos** [comment]
Alternatively `parse` could take an optional output parameter and return a pointer to the character past the last character parsed, similar to `strtof`. Given that as a basic building block, it would be easy to implement a convenience function (e.g. `parse_document` that takes that value and checks whether the remaining input `isspace` until the end and throws an appropriate exception if it is not.

Without this basic building block, users might find themselves in a situation where they have have a buffer with input (e.g. de-serializing JSON objects from a file), and they would like to extract the next valid JSON object, but they have to go through song-and-dance of creating `std::stringstream` just to be able to use `operator<<`.

**Comment by TurpentineDistillery** [comment]
@nlohmann 
I have already answered your question above: Ctrl+F for "There are two classes of use-cases:"

**Comment by nlohmann** [comment]
@TurpentineDistillery Thanks a lot. I should have gone through the thread again. Sorry.

Can you comment on @krzysztofwos remark on how to cope with reading a buffer?

**Comment by TurpentineDistillery** [comment]
On the lower level the parser should be implemented in terms of taking a pair of iterators `[begin, end)` (pointers into a char array, or istreambuf_iterator - whatever), and the `parse` and `operator<<` would be implemented on top of that.

**Comment by nlohmann** [comment]
Proposed fix in https://github.com/nlohmann/json/tree/feature/manual_lexer.

**Comment by nlohmann** [comment]
Now merged into develop. I shall update the behavior of `operator>>` and post an update here for final discussion.

**Comment by nlohmann** [comment]
Here is the current output:

```

Test-case[0]: ''
With int:
    final values: result=999 i=0 streampos=-1 streamflags=4098

With json:
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[1]: '   '
With int:
    final values: result=999 i=0 streampos=-1 streamflags=4098

With json:
    std::exception:[json.exception.parse_error.101] parse error at 4: syntax error - unexpected end of input

----

Test-case[2]: '111'
With int:
    result[0]=111; streampos=-1
    final values: result=111 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=111; streampos=3
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[3]: '222 	
'
With int:
    result[0]=222; streampos=3
    final values: result=222 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=222; streampos=6
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[4]: '
	 333'
With int:
    result[0]=333; streampos=-1
    final values: result=333 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=333; streampos=6
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[5]: ' 111 
222
 
  333'
With int:
    result[0]=111; streampos=4
    result[1]=222; streampos=9
    result[2]=333; streampos=-1
    final values: result=333 i=3 streampos=-1 streamflags=4098

With json:
    result[0]=111; streampos=9
    result[1]=333; streampos=17
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input
```

For some strange reason, the `222` in the last example is missed. All other inputs are processed as expected.

**Comment by nlohmann** [comment]
I understood the problem: after parsing the first number `111`, we read the next token `222`, but this token is not kept inside the parser, because it is destructed and reconstructed in the loop, so the next token that is processed is `333`.

**Comment by nlohmann** [comment]
This issue can be fixed by restructuring the parser. This is, however, not trivial and nothing I can do today. At the same time, issue #577 also requires a reimplementation of the parser, so we may have to wait a bit longer for this issue to close.

**Comment by ceztko** [comment]
I am adding the use case for #529 here, so that it can be tracked as well:

```
#include <fstream>
#include <sstream>
#include <json.hpp>

using namespace std;
using namespace nlohmann;

char* test =
"{\n\
    \"one\"   : 1,\n\
    \"two\"   : 2\n\
}\n\
{\n\
    \"three\" : 3\n\
}";

int main()
{
    string str(test);
    {
        ofstream file("test.json");
        file << str;
    }

    ifstream stream("test.json", ifstream::in);
    json val;
    try
    {
        size_t i = 0;
        while (stream.peek() != EOF)
        {
            stream >> val;
            cerr << "    result[" << i++ << "]=" << val
                << "; streampos=" << stream.tellg() << std::endl;
        }
    }
    catch (std::exception& e)
    {
        cerr << "    std::exception:" << e.what() << std::endl;
    }
    return 0;
}
```

With VS2017 it's failing in the second deserialization attempt, but the behavior seems to be dependent on caching.  If you intend to do some caching of the stream internally in json (and not read the stream character by character), then you may want to expose the internal parser. What I am trying to achieve is being able to parse the first two full roots elements in the first two iterations and be able to handle EOF correctly.

**Comment by nlohmann** [comment]
I did the restructuring I mentioned in https://github.com/nlohmann/json/issues/367#issuecomment-300528634. I did not find the time to test this issue again though.

**Comment by nlohmann** [comment]
The output for the example above (https://github.com/nlohmann/json/issues/367#issue-191383970) is now

```

Test-case[0]: ''
With int:
    final values: result=999 i=0 streampos=-1 streamflags=4098

With json:
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[1]: '   '
With int:
    final values: result=999 i=0 streampos=-1 streamflags=4098

With json:
    std::exception:[json.exception.parse_error.101] parse error at 4: syntax error - unexpected end of input

----

Test-case[2]: '111'
With int:
    result[0]=111; streampos=-1
    final values: result=111 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=111; streampos=3
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[3]: '222 	
'
With int:
    result[0]=222; streampos=3
    final values: result=222 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=222; streampos=3
    std::exception:[json.exception.parse_error.101] parse error at 4: syntax error - unexpected end of input

----

Test-case[4]: '
	 333'
With int:
    result[0]=333; streampos=-1
    final values: result=333 i=1 streampos=-1 streamflags=4098

With json:
    result[0]=333; streampos=6
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----

Test-case[5]: ' 111 
222
 
  333'
With int:
    result[0]=111; streampos=4
    result[1]=222; streampos=9
    result[2]=333; streampos=-1
    final values: result=333 i=3 streampos=-1 streamflags=4098

With json:
    result[0]=111; streampos=4
    result[1]=222; streampos=9
    result[2]=333; streampos=17
    std::exception:[json.exception.parse_error.101] parse error at 1: syntax error - unexpected end of input

----
Program ended with exit code: 0
```

This seems reasonable to me.

**Comment by nlohmann** [comment]
There was still a bug left, but now the examples from @TurpentineDistillery and @ceztko behave as expected. I added them as regression tests, see 9e507dfad59214483a20ffe34fde28b9b66ce6bb. If the CI build succeeds, I think I can close this issue.

**Comment by ceztko** [comment]
Hello, I just took the test case I posted here and executed it verbatim on VS2017.  Please note that the test is saving the stream to a file and after reading from it. The result is this:
```
    result[0]={"one":1,"two":2}; streampos=36
    std::exception:[json.exception.parse_error.101] parse error at 2: syntax error - unexpected '}'
```

After the first read the stream is actually postioned correctly after the first "}", but it's failing to read the second element. If I create a stringstream with exactly the same content with `stringstream stream(test);`, I get this result instead:
```
    result[0]={"one":1,"two":2}; streampos=36
    result[1]={"three":3}; streampos=56
```
Which means it correctly parsed the second element, without exceptions.  The stream is positioned correctly after the first "}" on the first read and before EOF on the second read. Behavior may be compiler/library dependent but I am expecting the same result to happen on the file stream. Question: are you parsing the stream character by character or are you doing some kind of caching?

**Comment by nlohmann** [comment]
@ceztko I also realized different behavior in MSVC as the tests on AppVeyor fail (https://ci.appveyor.com/project/nlohmann/json/build/1965). I haven't had the chance to check the logs though.

About caching: yes, I read 16 KB from an input stream into a cache. When the parser is destructed, I call `seekg` to the last actually processed character. I tested it with string streams and hoped this would also work with file streams.

**Comment by ceztko** [comment]
> I read 16 KB from an input stream into a cache. When the parser is
> destructed, I call seekg to the last actually processed character

It's perfectly reasonable to cache some amount of bytes in advance when reading from any stream. But I am not expecting the seekg to work reliably on any kind of stream (for example: if the stream is a socket). The idea should work on filestream, though, so there may be some bug. 
If you want to support streams that can be read forward-only, I recommend you to expose the internal json::parser class in the public api so any amount of unparsed cached data can't be lost.

**Comment by nlohmann** [comment]
In may add special overloads for ifstream and istringstream that use caches and remove caching for general istream cases. Would this work?

**Comment by ceztko** [comment]
> In may add special overloads for ifstream and istringstream that use caches
> and remove caching for general istream cases. Would this work?

For me it's also a workable solution. You may also consider having the caching only on ifstream since istringstream it's already on memory.

**Comment by nlohmann** [comment]
@ceztko I openend [a question at StackOverflow](https://stackoverflow.com/questions/44503211/how-to-correctly-read-from-already-opened-stdifstream-using-a-buffer) and found out the code I used to "rewind" the stream was not correct. I fixed it in a [feature branch](https://github.com/nlohmann/json/tree/feature/ifstream) and the example from https://github.com/nlohmann/json/issues/367#issuecomment-301788355 is now working with MSVC 2015 and MSVC 2017 (see [AppVeyor build](https://ci.appveyor.com/project/nlohmann/json)).

**Comment by ceztko** [comment]
@nlohmann great news, thanks! Sorry for not having looked at this: I wanted to debug it but I couldn't find the time.

**Comment by nlohmann** [comment]
Merged fd4a0ecabc301c5b6b551058080120d3bb6ee729 which fixes this issue. Thanks everybody for the patience.

**Comment by daniellicht13** [comment]
I see operator>> has been updated to parse multiple json objects from one stream. Has there been any comparable update for parsing strings? I am trying to access json objects stored as a c string in a shared memory segment (using mmap) and create an object for each one in an array of json objects. As of now my only success has been by storing the mmap data in a text file and then creating a stream from the text file to use the >> operator. I am reluctant to write my own parser for open and close brackets to divide up the text file as was suggested somewhere back in this thread or one of the related ones. Thanks
